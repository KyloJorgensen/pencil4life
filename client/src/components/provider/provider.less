
body {
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    padding: 0;
    margin: 0;
}

#app {
    position: inherit;
    top: 0;
    bottom: 0;
}

.provider {
    overflow: auto;
    height: 100%;
    width: 100%;
}