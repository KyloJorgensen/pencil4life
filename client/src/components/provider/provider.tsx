'use strict';

import * as React from 'react';

export const ProviderContext = React.createContext({
	addEventListener: (eventType:string, uniqueTag: string, eventHandler: Function):boolean => {
    console.error('Error: Cannot addEventListener no ProviderContext.Provider element in parents');
    return false;
  },
	removeEventListener: (eventType:string, uniqueTag:string):boolean => {
	  console.error('Error: Cannot removeEventListener no ProviderContext.Provider element in parents');
    return false;  
  },
});

export interface IEventListener {
  eventType: string,
  uniqueTag: string;
  eventHandler: Function;
}

export interface IProviderState {
  vaildEventTypes: string[];
  eventListeners: IEventListener[];
}

export class Provider extends React.Component<null, IProviderState> {
  constructor(props) {
    super(props);
    this.state = {
      vaildEventTypes: [
        'click'
      ],
      eventListeners: [],
    };
    this.handleClick = this.handleClick.bind(this);
    this.addEventListener = this.addEventListener.bind(this);
    this.removeEventListener = this.removeEventListener.bind(this);
  }

  handleClick = event => {
    this.state.eventListeners.filter(listner => listner.eventType == 'click').forEach(listener => listener.eventHandler(event));
  };

  addEventListener = (eventType:string, uniqueTag:string, eventHandler: Function):boolean => {
    let added = true;
    this.setState((prevState, props) => {
      const { eventListeners, vaildEventTypes } = prevState;

      if (!vaildEventTypes.includes(eventType)) {
        added = false;
      } else if ( eventListeners.filter(listener => listener.eventType == eventType).map(listener => listener.uniqueTag).includes(uniqueTag) ) {
        added = false;
      } else {
        eventListeners.push({
          eventType: eventType,
          uniqueTag: uniqueTag,
          eventHandler: eventHandler,
        })
      }
      return {eventListeners: eventListeners};
    });
    return added;
  }
  
	removeEventListener = (eventType:string, uniqueTag:string):boolean => {
    let deleted = true;
    this.setState((prevState, props) => {
      const { eventListeners } = prevState; 
      const index = eventListeners.findIndex(listener => listener.uniqueTag == uniqueTag && listener.eventType == eventType);
      if (index == -1) {
        deleted = false;
      } else {
        eventListeners.splice(index, 1);
      }
      return {eventListeners: eventListeners};
    });
    return deleted;
	}

  render() {
    const { addEventListener, removeEventListener, handleClick } = this;
		const context = {
			addEventListener: addEventListener,
			removeEventListener: removeEventListener,
    }
    
    return (
      <ProviderContext.Provider value={context}>
        <div onClick={handleClick}>
            {this.props.children}
        </div>
      </ProviderContext.Provider>
    );
  }
}

export default Provider 
