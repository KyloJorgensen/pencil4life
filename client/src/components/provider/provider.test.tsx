'use strict';

import * as React from 'react';
import { shallow, mount, ShallowWrapper, ReactWrapper } from 'enzyme';
import { Provider, ProviderContext} from './provider';

let wrapper: ShallowWrapper | ReactWrapper;

describe('Render ProviderContext', () => {
	it('matches the snapshot', () => {
		wrapper = mount(
			<ProviderContext.Consumer>
				{context => {
					return (
						<p></p>
					);
				}}
			</ProviderContext.Consumer>
		);
		expect(wrapper).toMatchSnapshot();
	});

	it('should have a p', () => {
		wrapper = mount(
			<ProviderContext.Consumer>
				{context => {
					return (
						<p></p>
					);
				}}
			</ProviderContext.Consumer>
		);
		const child = wrapper.find('p');
		expect(child.length).toEqual(1);
	});

	it('should have a p that contains "child"', () => {
		wrapper = mount(
			<ProviderContext.Consumer>
				{context => {
					return (
						<p>child</p>
					);
				}}
			</ProviderContext.Consumer>
		);
		const child = wrapper.find('p').contains('child');
		expect(child).toEqual(true);
	});
});

describe('Render Provider component', () => {
	it('matches the snapshot', () => {
		wrapper = shallow(<Provider />);
		expect(wrapper).toMatchSnapshot();
	});

	it('should have a p', () => {
		wrapper = shallow(
			<Provider>
				<p>child</p>
			</Provider>
		);
		const child = wrapper.find('p');
		expect(child.length).toEqual(1);
	});

	it('should have a p that contains "child"', () => {
		wrapper = shallow(
			<Provider>
				<p>child</p>
			</Provider>
		);
		const child = wrapper.find('p').contains('child');
		expect(child).toEqual(true);
	});
});

describe('interaction', () => {
	let uniqueNumber = 0;
	let getUniqueNumber = ():number => {
		uniqueNumber++;
		return uniqueNumber;
	};


	interface IInsideProps {
		addEventListener:Provider["addEventListener"];
		removeEventListener:Provider["removeEventListener"];
		mockHandler: Function;
	}	
	interface IInsideState {
		uniqueNumber: number;
	}
	class Inside extends React.Component<IInsideProps, IInsideState> {
		constructor(props) {
			super(props)
			this.state = {
				uniqueNumber: getUniqueNumber()
			}
		}

		componentDidMount() {
			this.props.addEventListener('click', this.state.uniqueNumber + 'uniqueTag', this.props.mockHandler);
		}

		render() {
			return (
				<p></p>
			);
		}
	}
	beforeEach(() => {
		uniqueNumber = 0;
	});

	it('should add event listener', () => {
		const mockHandler = jest.fn(() => {});
 		wrapper = mount(
			<Provider>
				<ProviderContext.Consumer>
					{context => {
						const props = {
							addEventListener: context.addEventListener,
							removeEventListener: context.removeEventListener,
							mockHandler: mockHandler
						}

						return (
							<Inside {...props} />
						);
					}}
				</ProviderContext.Consumer>
			</Provider>
		);

		wrapper.find('p').simulate('click');
		expect(mockHandler.mock.calls.length).toEqual(1);
	});

	it('should remove event listener', () => {
		const mockHandler = jest.fn(() => {});
		let props;
 		wrapper = mount(
			<Provider>
				<ProviderContext.Consumer>
					{context => {
						props = {
							addEventListener: context.addEventListener,
							removeEventListener: context.removeEventListener,
							mockHandler: mockHandler
						}

						return (
							<Inside {...props} />
						);
					}}
				</ProviderContext.Consumer>
			</Provider>
		);
		props.removeEventListener('click', '1uniqueTag');
		wrapper.find('p').simulate('click');
		expect(mockHandler.mock.calls.length).toEqual(0);
	});

	it('should not trigger on the outside', () => {
		const mockHandler = jest.fn(() => {});
 		wrapper = mount(
			<main>
				<Provider>
					<ProviderContext.Consumer>
						{context => {
							const props = {
								addEventListener: context.addEventListener,
								removeEventListener: context.removeEventListener,
								mockHandler: mockHandler
							}

							return (
								<Inside {...props} />
							);
						}}
					</ProviderContext.Consumer>
				</Provider>
			</main>
		);

		wrapper.find('main').simulate('click');
		expect(mockHandler.mock.calls.length).toEqual(0);
	});

	it('should trigger all "click" events on click', () => {
		const mockHandler = jest.fn(() => {});
 		wrapper = mount(
			<Provider>
				<ProviderContext.Consumer>
					{context => {
						const props = {
							addEventListener: context.addEventListener,
							removeEventListener: context.removeEventListener,
							mockHandler: mockHandler
						}

						return (
							<Inside {...props} />
						);
					}}
				</ProviderContext.Consumer>
				<ProviderContext.Consumer>
					{context => {
						const props = {
							addEventListener: context.addEventListener,
							removeEventListener: context.removeEventListener,
							mockHandler: mockHandler
						}

						return (
							<Inside {...props} />
						);
					}}
				</ProviderContext.Consumer>
				<ProviderContext.Consumer>
					{context => {
						const props = {
							addEventListener: context.addEventListener,
							removeEventListener: context.removeEventListener,
							mockHandler: mockHandler
						}

						return (
							<Inside {...props} />
						);
					}}
				</ProviderContext.Consumer>
			</Provider>
		);

		wrapper.find('p').first().simulate('click');
		expect(mockHandler.mock.calls.length).toEqual(3);
	});
});