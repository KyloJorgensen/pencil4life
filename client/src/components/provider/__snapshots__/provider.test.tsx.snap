// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Provider component matches the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <div
        onClick={[Function]}
      />,
      "value": Object {
        "addEventListener": [Function],
        "removeEventListener": [Function],
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": undefined,
        "onClick": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": "div",
    },
    "type": Object {
      "$$typeof": Symbol(react.provider),
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": [Circular],
        "_calculateChangedBits": null,
        "_changedBits": 0,
        "_changedBits2": 0,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": Object {
          "addEventListener": [Function],
          "removeEventListener": [Function],
        },
        "_currentValue2": Object {
          "addEventListener": [Function],
          "removeEventListener": [Function],
        },
        "_defaultValue": Object {
          "addEventListener": [Function],
          "removeEventListener": [Function],
        },
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <div
          onClick={[Function]}
        />,
        "value": Object {
          "addEventListener": [Function],
          "removeEventListener": [Function],
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": undefined,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": "div",
      },
      "type": Object {
        "$$typeof": Symbol(react.provider),
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": [Circular],
          "_calculateChangedBits": null,
          "_changedBits": 0,
          "_changedBits2": 0,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": Object {
            "addEventListener": [Function],
            "removeEventListener": [Function],
          },
          "_currentValue2": Object {
            "addEventListener": [Function],
            "removeEventListener": [Function],
          },
          "_defaultValue": Object {
            "addEventListener": [Function],
            "removeEventListener": [Function],
          },
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Render ProviderContext matches the snapshot 1`] = `
ReactWrapper {
  "length": 1,
  Symbol(enzyme.__unrendered__): <Context.Consumer>
    [Function]
  </Context.Consumer>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__node__): Object {
    "instance": <p />,
    "key": undefined,
    "nodeType": "host",
    "props": Object {},
    "ref": null,
    "rendered": Array [
      undefined,
    ],
    "type": "p",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": <p />,
      "key": undefined,
      "nodeType": "host",
      "props": Object {},
      "ref": null,
      "rendered": Array [
        undefined,
      ],
      "type": "p",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
